src/main/kotlin/org/rust/ide/surroundWith/statement/RustStatementsSurrounderBase.kt

```diff
@@ -1,24 +1,44 @@
 package org.rust.ide.surroundWith.statement
 
+import com.intellij.codeInsight.CodeInsightUtilBase
 import com.intellij.lang.surroundWith.Surrounder
 import com.intellij.openapi.editor.Editor
 import com.intellij.openapi.project.Project
 import com.intellij.openapi.util.TextRange
 import com.intellij.psi.PsiElement
+import org.rust.ide.surroundWith.addStatements
+import org.rust.lang.core.psi.RustBlockElement
 
 abstract class RustStatementsSurrounderBase : Surrounder {
-    abstract fun surroundStatements(
-        project: Project,
-        editor: Editor,
-        container: PsiElement,
-        statements: Array<out PsiElement>
-    ): TextRange?
+    abstract fun createTemplate(project: Project): PsiElement
+    abstract fun getCodeBlock(expression: PsiElement): RustBlockElement
 
-    final override fun isApplicable(elements: Array<out PsiElement>): Boolean = elements.isNotEmpty()
+    open fun getExprForRemove(expression: PsiElement): PsiElement? = null
+    open fun isNeedRemoveExpr() = false
```

Comment:
There functions can be `protected abstract` and `protected open`. `isNeedRemoveExpr` could be a `val`:

```
protected open val shouldRemoveExpr: Boolean
```

Even better, I think you can have only `getExprForRemove`, which would return `null` if nothing should be removed.